### PR TITLE
[SOFT-171] Implement bms relay control

### PR DESCRIPTION
**Pull request by g-jessmuir**
Implements the BMS relay sequence

### CODE REVIEW

File: projects/bms_carrier/src/relay_sequence.c [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+void prv_io_int_callback(const GpioAddress *address, void *context) {
+  RelayStorage *storage = context;
+  Mcp23008GpioState hv_state = NUM_MCP23008_GPIO_STATES;
+  Mcp23008GpioState gnd_state = NUM_MCP23008_GPIO_STATES;
+  mcp23008_gpio_get_state(&s_mcp23008_hv, &hv_state);
+  mcp23008_gpio_get_state(&s_mcp23008_gnd, &gnd_state);
+  storage->gnd_enabled = gnd_state == MCP23008_GPIO_STATE_HIGH ? true : false;
+  storage->hv_enabled = hv_state == MCP23008_GPIO_STATE_HIGH ? true : false;
```
[the remark this answers is not in the document]
Comment (g-jessmuir) [reply projects/bms_carrier/src/relay_sequence.c]: My bad, fixed

File: projects/bms_carrier/src/relay_sequence.c [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  } else if (msg->msg_id == SYSTEM_CAN_MESSAGE_POWER_OFF_SEQUENCE) {
+    uint16_t step = NUM_EE_POWER_OFF_SEQUENCES;
+    CAN_UNPACK_POWER_OFF_SEQUENCE(msg, &step);
+    if (step == EE_POWER_OFF_SEQUENCE_OPEN_BATTERY_RELAYS) {
+      relay_open_sequence(storage);
+    }
+  }
+  *ack = CAN_ACK_STATUS_OK;
+  return STATUS_CODE_OK;
```
[the remark this answers is not in the document]
Comment (g-jessmuir) [reply projects/bms_carrier/src/relay_sequence.c]: I dunno what I was thinking, moved to separate `prv_power_on/off_rx` respectively

File: projects/bms_carrier/test/test_relay_sequence.c [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    s_hv_state = state;
+  } else if (address->pin == 1) {
+    s_gnd_state = state;
+  }
+  return STATUS_CODE_OK;
+}
+
+static uint8_t s_fault_bps_bitmask = 0;
+static uint8_t s_fault_bps_calls = 0;
+void TEST_MOCK(fault_bps)(uint8_t bitmask, bool clear) {
```
[the remark this answers is not in the document]
Comment (g-jessmuir) [reply projects/bms_carrier/test/test_relay_sequence.c:56]: true, added assertion

File: projects/bms_carrier/test/test_relay_sequence.c [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  TEST_ASSERT_EQUAL(MCP23008_GPIO_STATE_LOW, s_storage.gnd_enabled);
+  TEST_ASSERT_EQUAL(MCP23008_GPIO_STATE_LOW, s_storage.hv_enabled);
+  s_23008_state_0 = MCP23008_GPIO_STATE_HIGH;
+  s_23008_state_1 = MCP23008_GPIO_STATE_HIGH;
+  gpio_it_trigger_interrupt(&s_io_expander_int);
+  TEST_ASSERT_EQUAL(MCP23008_GPIO_STATE_HIGH, s_storage.gnd_enabled);
+  TEST_ASSERT_EQUAL(MCP23008_GPIO_STATE_HIGH, s_storage.hv_enabled);
+}
+
+void test_open_good(void) {
```
[the remark this answers is not in the document]
Comment (g-jessmuir) [reply projects/bms_carrier/test/test_relay_sequence.c:94]: I found the soft timer assertions helped catch logic errors while I was writing the module. Do you think it's too verbose if I leave it in?

File: projects/bms_carrier/test/test_relay_sequence.c [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+  // no faults should have occured
+  TEST_ASSERT_EQUAL(0, s_fault_bps_calls);
+  TEST_ASSERT_EQUAL(0, s_fault_bps_bitmask);
+  delay_ms(RELAY_SEQUENCE_NEXT_STEP_DELAY_MS - RELAY_SEQUENCE_ASSERTION_DELAY_MS + 5);
+  // no timers should be running
+  TEST_ASSERT_EQUAL(0, soft_timer_remaining_time(s_storage.assertion_timer_id));
+  TEST_ASSERT_EQUAL(0, soft_timer_remaining_time(s_storage.next_step_timer_id));
+  // state should be not enabled
+  TEST_ASSERT_EQUAL(false, s_storage.hv_enabled);
+  TEST_ASSERT_EQUAL(false, s_storage.gnd_enabled);
```
[the remark this answers is not in the document]
Comment (g-jessmuir) [reply projects/bms_carrier/test/test_relay_sequence.c]: True, replaced a few things with `TEST_ASSERT_TRUE/FALSE` as necessary.

File: projects/bms_carrier/test/test_relay_sequence.c [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+  can_init(&s_can_storage, &s_can_settings);
+}
+
+void teardown_test(void) {}
+
+void test_io_int(void) {
+  TEST_ASSERT_OK(relay_sequence_init(&s_storage));
+  // test storage is set correctly
+  TEST_ASSERT_EQUAL(MCP23008_GPIO_STATE_LOW, s_storage.gnd_enabled);
+  TEST_ASSERT_EQUAL(MCP23008_GPIO_STATE_LOW, s_storage.hv_enabled);
```
[the remark this answers is not in the document]
Comment (g-jessmuir) [reply projects/bms_carrier/test/test_relay_sequence.c]: My bad, changed these to `TEST_ASSERT_TRUE`

File: projects/bms_carrier/test/test_relay_sequence.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (GeraldNDA) [reply projects/bms_carrier/test/test_relay_sequence.c:94]: That's fair, but I think it's an implementation detail that can change over time. The tests should verify that it works as in it responds as expected to various inputs/uses. IMO if someone were to restructure it significantly, the test should still work if the interface is the same.

[the remark this answers is not in the document]
Comment (g-jessmuir) [reply projects/bms_carrier/test/test_relay_sequence.c:94]: Actually when you put it that way that makes a lot of sense, I took out those assertions.

**Pull request merged**